Consolidated changes to support call_tree API

amarathe84 opened the pull request:
PR to support call_tree API. Not ready for formal review, but early feedback welcome!
- Separated call_tree from the replay utility. 
- Added example test cases for call_tree. 
- Integrated with the replay utility

rayandrew commented on examples/call_tree_example1.cpp:
```diff
@@ -0,0 +1,90 @@
+
+int main(int argc, char* argv[]) {
+    std::cout << "=== CallTree API Example 1: Basic Usage ===" << std::endl;
+    std::cout << std::endl;
```
For next dftracer-utils release, I think we should not use `std::cout` anymore and probably use `printf`/`std::printf` or `DFTRACER_UTILS_LOG_*`

Please use `printf` for now

rayandrew commented on include/dftracer/utils/utilities/call_tree/call_tree_internal.h:
can we separate this file to `include/dftracer/utils/utilities/call_tree/call_tree/internal/factory.h` and so on?

I believe it will enhance searchability and debuggability.

rayandrew commented on include/dftracer/utils/utilities/hash/hasher_utility.h:
any reason why we delete xxhash @amarathe84 ?

Let me know if you encounter any problems with the hash.

However, I agree to make the default as `std::hash`.

rayandrew commented on src/dftracer/utils/utilities/call_tree/call_tree.cpp:
```diff
@@ -0,0 +1,784 @@
+    bool found = impl_->find_trace_files(trace_dir, pattern);
+
+    if (found) {
+        std::cout << "Found " << impl_->trace_files.size() << " trace files in "
```
use `DFTRACER_UTILS_LOG_*` in the library

hariharan-devarajan commented on include/dftracer/utils/utilities/call_tree/call_tree_mpi.h:
@rayandrew The reason we did it in DLIO is that we may change between MPI and DDP or something else. Here let's keep it simple. The future expected model is MPI + threads and so this should be fine. Also in MPI +X paradigms making a singleton of rank can give u a a world of pain :D. So let's keep them as it is.

amarathe84 commented on src/CMakeLists.txt:
Updated CMakeLists.

amarathe84 replied:
Need to add documentation RST. Will work on it shortly.

amarathe84 commented:
> hi @amarathe84 can we fix the CI error? we need to remove XXHash tests since it will be removed

Yes,my last commit broke CI--working on fixing it now

amarathe84 committed:
Add call-tree utility with MPI support, documentation, and CI tests

- Implement MPICallTreeBuilder for distributed call graph generation
- Add internal call tree data structures (CallTree, ProcessCallTree, CallNode)
- Support for building hierarchical call graphs from DFTracer traces
- MPI-parallel processing with PID distribution across ranks
- Multiple output formats: binary, JSON (Chrome Tracing), and text
- Add comprehensive documentation (call-tree.rst)
- Add CI tests for core functionality, configuration, and serialization
- Fix code formatting issues to pass clang-format checks

amarathe84 committed:
Fix CallTree test: get() returns nullptr for non-existent keys

The test incorrectly expected get() to return non-null for a key that
doesn't exist yet. Fixed by:
1. First checking get() returns nullptr for non-existent keys
2. Using operator[] to create the entry
3. Then verifying get() returns the created entry
